### PR TITLE
Implement Reusable Gatsby Theme

### LINKED ISSUE #3114: Extract docs engine into reusable package
We currently use our custom docs engine on dvc.org and cml.dev, with plans to add it to all our product sites. The docs engine is currently "shared" between the two sites by simply copy-pasting 90+% of the code between the dvc.org and cml.dev repos. This already poses a problem when it comes to updating the engine, and will only get worse as we introduce new sites.

Due to these requirements and issues, we're undertaking an effort to move the docs engine to a reusable package, leaning on the Gatsby Plugin and Theme API as well as more general Node package interfaces like module exports. The long-term plan is to package the docs engine as an NPM package.

To ease the transition, we'll break development into multiple phases

### Phase 1: **dvc.org local plugin**
We'll start by moving all dvc.org docs logic into a local Gatsby plugin, a pseudo-npm package in the `plugins` folder. This will allow for much easier rapid incremental development with minimal worry about maintaining updates for a half-working package. This phase ends after we move all of `dvc.org`'s docs engine to this local package.

- [x] #3113 
- [x] Move docs sidebar logic to local plugin (https://github.com/iterative/dvc.org/pull/3170)
- [x] Move generalized React components to local plugin (either as exports or shadowable components)(https://github.com/iterative/dvc.org/pull/3170)

### Phase 2: **Internal package sharing**  
Once we're confident we've moved all the docs code we want to share, we can start turning the local plugin into a real package and use it among all our internal sites. This phase specifically puts less of an emphasis on how external users would use the plugin, instead putting the most consideration on how we'll use the plugin for our product sites.

- [x] Create a new repo and NPM package with the contents of the local plugin folder(https://github.com/iterative/gatsby-theme-iterative/)
- [ ] Remove the local plugin from dvc.org and ensure all imports are changed to the npm package name (https://github.com/iterative/dvc.org/pull/3396)
- [x] Remove the copy-paste docs engine from cml.dev and use the reusable package(https://github.com/iterative/cml.dev/pull/233)

### Phase 3: **Generalizing for outside audiences**
After we're comfortably using this package for all our internal websites, we can consider the possibility of sharing it with the public as an option for other Gatsby sites to use as a docs engine. This is easily the most variable step- it could range from a massive refactor to being scrapped altogether if we decide to only ever consider internal use.

### LINKED ISSUE #3589: `cli` highlighter (update `dvc`)
[mlem.ai has](https://github.com/iterative/mlem.ai/issues/59) a `cli` Prism language that includes `dvc`, `cml`, and `mlem` commands (along with regular `shell` stuff).

- [x] Can we get it ported here @iterative/websites ?

_Originally posted by @jorgeorpinel in https://github.com/iterative/dvc.org/pull/3577#discussion_r881118100_

Also,

- [ ] Update existing `dvc` blocks? At least those that use other commands, from Linux tools to CML.

### PR DESCRIPTION
**Pull request by rogermparent**
This PR is made to try out `dvc.org` against the new shared website theme: [`@dvcorg/gatsby-theme-iterative`](https://github.com/iterative/gatsby-theme-iterative)! The parts of `dvc.org` that are expected to be used across multiple websites are removed in this PR and imported from the theme. 

Things this PR does:
- Removes the local plugin version of the theme and the glue code that made it 
- Removes components and logic that are now fully handled by the theme
- Moves components and logic that shadow (replace) files in the theme into the relevant shadowing directory, "src/@dvcorg/gatsby-theme-iterative".
- Updates packages (figured that making a reusable library with pinned out-of-date packages would be an issue, but possibly could have been skipped) and [makes refactors to fix what that breaks](https://github.com/iterative/dvc.org/pull/3396#discussion_r844617849)

- Closes #3589
- Closes #3114

### CODE REVIEW

File: src/components/Blog/Post/Markdown/index.tsx
Comment (julieg18) [inline src/components/Blog/Post/Markdown/index.tsx:1]: Not sure why, but `external-link` elements don't seem to be working: 

image

File: src/templates/doc-jsx.tsx
Comment (julieg18) [inline src/templates/doc-jsx.tsx:7]: External links icons aren't the right icon:

image

File: src/pages/doc/user-guide/glossary.tsx
Comment (julieg18) [inline src/pages/doc/user-guide/glossary.tsx:1]: Basic concepts pages are giving 404s. For example, the route [`doc/user-guide/basic-concepts/dvc-project`](https://dvc.org/doc/user-guide/basic-concepts/dvc-project) is leading to a 404 though it should exist 🤔 

https://dvc-org-try-external-th-doe8zu.herokuapp.com/doc/user-guide/basic-concepts/dvc-project

**Review by julieg18 — COMMENTED** [review]
Great work so far! Looking forward to seeing this in action for other websites!

File: package.json
Comment (yathomasi) [reply package.json:40]: Yes, https://github.com/iterative/gatsby-theme-iterative must be the one. It should be accessible once permission is given.
And, it's already published https://www.npmjs.com/package/@dvcorg/gatsby-theme-iterative.

File: src/pages/doc/user-guide/glossary.tsx
Comment (yathomasi) [reply src/pages/doc/user-guide/glossary.tsx:1]: The glossary page seems to be generated from basic concepts. So, that must be okay. I can see it on https://dvc-org-try-external-th-doe8zu.herokuapp.com/doc/user-guide/glossary

File: src/templates/doc-jsx.tsx
Comment (rogermparent) [reply src/templates/doc-jsx.tsx:7]: Good catch, I had to shuffle around some icons to make them importable from the theme and this icon was one of the ones changed most often.

File: src/pages/doc/user-guide/glossary.tsx
Comment (rogermparent) [reply src/pages/doc/user-guide/glossary.tsx:1]: Totally forgot we made these pages, I think the only thing stopping these pages from being made is an early return in the theme code that I put in thinking about the site before we made these into pages. Thanks!

File: src/components/Blog/Search/SearchBox/styles.module.css
Comment (rogermparent) [reply src/components/Blog/Search/SearchBox/styles.module.css:23]: Just ran `stylelint` and it replaced everything with double quotes, not actually sure why single quotes were committed but my guess is that I just hadn't run `stylelint` at the point.

File: package.json
Comment (rogermparent) [reply package.json:40]: Just realized that permission isn't on by default, I added @iterative/websites to the "Maintainer" role so we should all be able to see it now!

File: src/pages/doc/user-guide/glossary.tsx
Comment (rogermparent) [reply src/pages/doc/user-guide/glossary.tsx:1]: Fixed these with a published change to the theme. Maybe we can make it configurable in the future, but we won't use it internally so we can cross that bridge when we come to it later.

File: src/components/Blog/Post/Markdown/index.tsx
Comment (rogermparent) [reply src/components/Blog/Post/Markdown/index.tsx:1]: These got scooped up into the theme, but the plugin's still added into the remark definition so it's unfortunately not as easy as the plugin not being added. I'll have to look into it more!

Comment (rogermparent) [reply src/components/Blog/Post/Markdown/index.tsx:1]: [Fixed![(https://dvc-org-try-external-th-doe8zu.herokuapp.com/blog/march-22-heartbeat#march-office-hours)

Basically, the issue stems from me upgrading all our packages in this PR (probably unnecessary looking back on it, but we'd have to eventually anyway)

Most packages in the `unified` family like `remark`, `rehype` and derivatives were changed to only export ESM modules, preventing us from importing them with `require` in non-webpacked parts of the application like `gatsby-node` and `gatsby-config` (TypeScript support didn't help here unfortunately).

What does work is dynamic `import` calls, but that returns `Promise`s and the `import` calls can't be at the top level; by extension, all functions that consume them must be turned asynchronous.

And that's what I did! Our custom remark plugins (these elements as well as the image resizing plugin) are now all async functions which use dynamic imports for the packages that need them. Took a bit of refactoring and I had some stumbles along the way, but it seems they work just as well as before now and are more future-proofed.

TL;DR: updating remark broke our remark plugins, and I just finished up a moderately large refactor that was needed to get them working again.

File: src/templates/doc-jsx.tsx
Comment (rogermparent) [reply src/templates/doc-jsx.tsx:7]: [Fixed!](https://dvc-org-try-external-th-doe8zu.herokuapp.com/doc/user-guide/running-dvc-on-windows)
image


The files in question are `external-link-icon.svg` (the big gray one here) and `external-link.svg` (the correct small blue one). Better names might be in order here, but aside from that I checked for uses of `external-link.svg` in master and found this is the only place that icon is used, and I'm pretty sure we only have two, so this fix should take care of this kind of problem.

File: src/@dvcorg/gatsby-theme-iterative/components/HamburgerMenu/styles.module.css
Comment (rogermparent) [inline src/@dvcorg/gatsby-theme-iterative/components/HamburgerMenu/styles.module.css:138]: `--font-brandon` is now `--font-base` to be more generic in the theme, which trickles down here in its implementation.

File: src/@dvcorg/gatsby-theme-iterative/components/LayoutFooter/index.tsx
Comment (rogermparent) [inline src/@dvcorg/gatsby-theme-iterative/components/LayoutFooter/index.tsx:19]: Tons of things needed to be bumped two directories up because of the theme shadowing dir

File: src/@dvcorg/gatsby-theme-iterative/components/LayoutFooter/index.tsx
Comment (rogermparent) [inline src/@dvcorg/gatsby-theme-iterative/components/LayoutFooter/index.tsx:15]: Changed theme package name touches a lot of lines as well

File: src/@dvcorg/gatsby-theme-iterative/components/Page/getLayoutComponent.ts
Comment (rogermparent) [inline src/@dvcorg/gatsby-theme-iterative/components/Page/getLayoutComponent.ts:26]: This broken out layout logic file lets us reuse our weird layout system with whatever components or logic we want with minimal need for shadowing- the only reason we need it here is the blog layout, sites without a blog likely won't need to shadow this at all!

That said, this kind of system that imports all the components into one super-component is generally considered a bad idea for code splitting reasons. Obviously we did it for a reason, but I think looking into a better solution is warranted in the future.

File: src/@dvcorg/gatsby-theme-iterative/getEditLink.ts
Comment (rogermparent) [inline src/@dvcorg/gatsby-theme-iterative/getEditLink.ts:3]: This is what broke edit links last time. `constants` logic was changed to a simple function override- does the same thing, adaptable to any format we want, practically same amount of effort for changes.

File: src/@dvcorg/gatsby-theme-iterative/redirects.js
Comment (rogermparent) [inline src/@dvcorg/gatsby-theme-iterative/redirects.js:1]: Must be exported from here to shadow, the file contents could also live here (and be my preference) but I figured I'd do what I can to minimize the amount of unnecessary changes tossed onto this already massive pile.

File: src/components/Blog/Feed/Item/index.tsx
Comment (rogermparent) [inline src/components/Blog/Feed/Item/index.tsx:117]: These `picture` GraphQL endpoints now point to `File` nodes instead of `ImageSharp` nodes, using a simpler query method under the hood as well as paving the way for supporting files `sharp` doesn't support, like `SVG`s.

File: src/components/Community/Section/styles.module.css
Comment (rogermparent) [inline src/components/Community/Section/styles.module.css:64]: This was a stylelint thing and it seems to [be supported in the CSS colors spec v4](https://www.w3.org/TR/css-color-4/#currentcolor-color), but I'm not 100% sure this doesn't break things in older browsers.

File: src/components/Home/WhatsNewModal/styles.module.css
Comment (rogermparent) [inline src/components/Home/WhatsNewModal/styles.module.css:73]: inconsistent linter weirdness

File: src/gatsby/common.js
Comment (rogermparent) [inline src/gatsby/common.js:19]: Weird async remark import stuff

File: src/gatsby/models/authors/createSchemaCustomization.js
Comment (rogermparent) [inline src/gatsby/models/authors/createSchemaCustomization.js:42]: Simpler, closer to built-in query for linking avatars/pictures to files which doesn't need a whole module dedicated to it.

File: tsconfig.json
Comment (rogermparent) [inline tsconfig.json:22]: We never emit our typescript so it may as well be in our config.

Also stops the vscode linter from yelling at me for what otherwise would be invalid tsconfig.

**Review by rogermparent — COMMENTED** [review]
I believe all the current issues pointed out have been fixed, and I think we're ready for hopefully a final round of reviews!

This PR is **big**, and a lot of it is import changes, so I tried my best to point out everything important with line comments.

...Oh, the ts linter found a lot of issues, I guess if everything else works that'll be the final task.

**Comment by rogermparent** [comment]
Alright, now all the lint issues are cleared and I think we're ready for another round of reviews!

One thing I noticed: the same image on the March Community Gems post shows up smaller on dvc.org compared to the deploy preview:
image

Not sure why, there's no resize instruction so I don't think the resize plugin should be affecting either version's images.

**Review by julieg18 — COMMENTED** [review]
Great job! Not seeing any issues!

> One thing I noticed: the same image on the March Community Gems post shows up smaller on dvc.org compared to the deploy preview:
> Not sure why, there's no resize instruction so I don't think the resize plugin should be affecting either version's images.

Weird! The only difference appears to be a difference in `max-width`, which is generated in our `resize-image-plugin`.  Since its our own code and not a external package, should be simple enough to debug 🤔

**Comment by yathomasi** [comment]
Since gatsby's build was failing, I tried deploying it on [Heroku](https://dvcorg-theme.herokuapp.com/).

**Review by yathomasi — APPROVED** [review]
Great job @rogermparent.
Besides the difference in the image already mentioned I didn't see any major issue.

I think this is ready for merge as we already have merged the docs-engine theme on other websites. It will also catch up with added features like `cli` language implementation on the docs-engine theme, once this PR gets merged.

@iterative/websites please do give a final review.

File: src/components/Home/Diagram/index.tsx
Comment (julieg18) [inline src/components/Home/Diagram/index.tsx:5]: Text isn't arranged right in the desktop version: 

image

**Review by julieg18 — CHANGES_REQUESTED** [review]
I vote we at least fix the error in the diagram section before merging :) We probably also should open an issue about the rest of the minor things so we don't forget to fix them later!

File: src/components/Home/Diagram/index.tsx
Comment (rogermparent) [reply src/components/Home/Diagram/index.tsx:5]: I don't get the issue on Firefox, but I do on Chrome. Seems to just be a simple flexbox issue because the columns are so close to 100%- if you set `box-sizing: border-box` on the `column` class it resolves the issue.

Comment (rogermparent) [reply src/components/Home/Diagram/index.tsx:5]: Ah yeah, and the rule is already there in the stylesheet but not applying because of some weird postcss stuff

Comment (rogermparent) [reply src/components/Home/Diagram/index.tsx:5]: By using https://github.com/iterative/gatsby-theme-iterative/pull/20 to disable tailwind, the issue is fixed. I think it can serve as a band-aid while we try to get a more real fix in.

Comment (yathomasi) [reply src/components/Home/Diagram/index.tsx:5]: > I don't get the issue on Firefox, but I do on Chrome. Seems to just be a simple flexbox issue because the columns are so close to 100%- if you set `box-sizing: border-box` on the `column` class it resolves the issue.

Yes, it could be solved by overriding the CSS.

> Ah yeah, and the rule is already there in the stylesheet but not applying because of some weird postcss stuff

But, yes the real issue was with `postcss-nested` not working properly.

**Comment by yathomasi** [comment]
@rogermparent Thanks for the fix.
Deployed to https://dvcorg-theme.herokuapp.com.

Not related to this PR, did we purposefully not use Gatsby's Static Image? (reason could be image url) 
Screen Shot 2022-05-27 at 12 20 29

Screen Shot 2022-05-27 at 12 20 20

**Review by julieg18 — COMMENTED** [review]
Site looks good to merge! But what exactly is breaking Gatsby Cloud? Once this is merged, will the preview break on any pr that uses the latest version of master? If so, we may want to solve what's breaking Gatsby Cloud or wait to merge this until automatic Heroku deployments are back up 🤔

**Comment by yathomasi** [comment]
Thanks, @julieg18 for the farsighted review.

> we may want to solve what's breaking Gatsby Cloud or wait to merge this until automatic Heroku deployments are back up

Yes, you are right.

**Comment by rogermparent** [comment]
> Not related to this PR, did we purposefully not use Gatsby's Static Image?

I don't know for sure since that page was built before I was on the team, but my guess it that it's a holdover from Nextjs / early Gatsby transition that we never bothered to change. Seems like we would only stand to gain from changing them to `gatsby-plugin-image` images.

**Comment by jorgeorpinel** [comment]
- [ ] Maybe we should move some of [these issues](https://github.com/iterative/dvc.org/issues?q=is%3Aopen+is%3Aissue+label%3A%22website%3A+eng-doc%22) to the Gatsby theme repo when this is merged.

**Comment by yathomasi** [comment]
> * [ ]  Hi. Maybe we should move some of [these issues](https://github.com/iterative/dvc.org/issues?q=is%3Aopen+is%3Aissue+label%3A%22website%3A+eng-doc%22) to the Gatsby theme repo when this is merged.

I can see most of those issues fit the theme repo. We can do that gradually. We can open new issues on theme repo and solving or moving these when we look into them.

**Comment by yathomasi** [comment]
The base reason for failing this PR against link-checker seems to be due to huge diff (https://github.com/iterative/dvc.org/pull/3625#issue-1262493226). Let's create new issue or move discussion to already existing link-checker issue(maybe https://github.com/iterative/dvc.org/issues/1838). 

And, minor issues are addressed and also we can create issue for that later as well. So, let's merge this PR unless you guys see any other issue. 
cc: @rogermparent @julieg18

File: plugins/gatsby-theme-iterative-docs/src/components/Documentation/Layout/SearchForm/index.tsx
Comment (julieg18) [reply plugins/gatsby-theme-iterative-docs/src/components/Documentation/Layout/SearchForm/index.tsx:47]: It's also too wide, but we could probably fix that on another iteration.

**Review by julieg18 — COMMENTED** [review]
Good to hear that the only issue with the link checker is that `diff` is too big. 

With the pr not breaking any new pr's link-check action, we should be able to merge once we stop the doc search from being disabled.

**Review by julieg18 — APPROVED** [review]
Fantastic work @yathomasi and @rogermparent 👏 👏 👏

**Pull request merged**
